### PR TITLE
[Merged by Bors] - feat(MorphismProperty/Presheaf): Define relatively representable morphisms

### PR DESCRIPTION
**Pull request by callesonne**
This PR defines relatively representable morphisms, and adds some basic API.

Co-authored-by: Joël Riou <rioujoel@gmail.com>

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/CategoryTheory/MorphismProperty/Presheaf.lean
Comment (joelriou) [inline Mathlib/CategoryTheory/MorphismProperty/Presheaf.lean:103]:
```suggestion
  (hf'.yoneda_map_fst' _) ▸ hf'.isPullback g
```

File: Mathlib/CategoryTheory/MorphismProperty/Presheaf.lean
Comment (joelriou) [inline Mathlib/CategoryTheory/MorphismProperty/Presheaf.lean:159]:
```suggestion
variable {Z : C} (i : Z ⟶ Y) (h : Z ⟶ X) (hi : yoneda.map i ≫ f' = yoneda.map h ≫ g)
```

File: Mathlib/CategoryTheory/MorphismProperty/Presheaf.lean
Comment (dagurtomas) [inline Mathlib/CategoryTheory/MorphismProperty/Presheaf.lean:18]:
```suggestion
* `Presheaf.representable` is a `MorphismProperty` expressing the fact that a morphism `f` of presheaves
  is representable, i.e. that any morphism from a representable presheaf to the target can be pulled back
  along `f` to a morphism from a representable presheaf to the source.
```
Feel free to rephrase, I just think there should be a short description here of what this means mathematically.

File: Mathlib/CategoryTheory/MorphismProperty/Presheaf.lean
Comment (dagurtomas) [inline Mathlib/CategoryTheory/MorphismProperty/Presheaf.lean:28]:
```suggestion
*  Whenever `f` is of type `yoneda.obj Y ⟶ G`, we also have `hf.fst' g : hf.pullback g ⟶ Y`
which is the preimage under `yoneda` of `hf.fst g`.
```

File: Mathlib/CategoryTheory/MorphismProperty/Presheaf.lean
Comment (dagurtomas) [inline Mathlib/CategoryTheory/MorphismProperty/Presheaf.lean:30]:
```suggestion
* `hom_ext`, `hom_ext'`, `lift`, `lift'` are variants of the universal property of
  `yoneda.obj (hf.pullback g)`, where as much as possible has been formulated internally to `C`.
```

File: Mathlib/CategoryTheory/MorphismProperty/Presheaf.lean
Comment (dagurtomas) [inline Mathlib/CategoryTheory/MorphismProperty/Presheaf.lean:34]: Is there a reason for this blank line between bullet points? If not, could you remove it?

File: Mathlib/CategoryTheory/MorphismProperty/Presheaf.lean
Comment (dagurtomas) [inline Mathlib/CategoryTheory/MorphismProperty/Presheaf.lean:152]:
```suggestion

@[reassoc (attr := simp)]
lemma lift_snd : hf.lift i h hi ≫ hf.snd g = h :=
```

**Review by dagurtomas — APPROVED** [review]

File: Mathlib/CategoryTheory/MorphismProperty/Presheaf.lean
Comment (callesonne) [reply Mathlib/CategoryTheory/MorphismProperty/Presheaf.lean:18]: Good suggestion, I think I will spell it out a bit more even.

File: Mathlib/CategoryTheory/MorphismProperty/Presheaf.lean
Comment (callesonne) [reply Mathlib/CategoryTheory/MorphismProperty/Presheaf.lean:34]: I think in my head these blank lines separated different categories somehow, but now that I look back on it I think that no blank line is better

**Comment by dagurtomas** [comment]
Thanks!

maintainer merge

File: Mathlib/CategoryTheory/MorphismProperty/Presheaf.lean
Comment (semorrison) [inline Mathlib/CategoryTheory/MorphismProperty/Presheaf.lean:66]: I wonder if this should be generalized, i.e. let `Cᵒᵖ ⥤ Type v` just be any category, and `yoneda` just be any functor into it. The definition makes sense, and also a reasonable amount of the API below.

Comment (semorrison) [reply Mathlib/CategoryTheory/MorphismProperty/Presheaf.lean:66]: (And then have a specialized version for `yoneda`.)

Comment (semorrison) [reply Mathlib/CategoryTheory/MorphismProperty/Presheaf.lean:66]: `hasRelativePullbacks`, perhaps?

Comment (semorrison) [reply Mathlib/CategoryTheory/MorphismProperty/Presheaf.lean:66]: (I don't intend this comment to block merging, if it doesn't sounds like a good idea!)

Comment (callesonne) [reply Mathlib/CategoryTheory/MorphismProperty/Presheaf.lean:66]: I think this sound interesting / fun, so I will try it

Comment (callesonne) [reply Mathlib/CategoryTheory/MorphismProperty/Presheaf.lean:66]: Actually, maybe it's not so bad.

Comment (callesonne) [reply Mathlib/CategoryTheory/MorphismProperty/Presheaf.lean]: Done!

File: Mathlib/CategoryTheory/MorphismProperty/Representable.lean
Comment (semorrison) [inline Mathlib/CategoryTheory/MorphismProperty/Representable.lean:35]:
```suggestion
* We define `relativelyRepresentable` as a `MorphismProperty`. A morphism `f : X ⟶ Y` in `D` is
```

File: Mathlib/CategoryTheory/MorphismProperty/Representable.lean
Comment (semorrison) [inline Mathlib/CategoryTheory/MorphismProperty/Representable.lean:36]:
```suggestion
  said to be relatively representable with respect to `F` if for any `g : F.obj a ⟶ Y`, there exists a pullback square
```

**Comment by semorrison** [comment]
bors merge

**Pull request closed without merging**
